pints/_transformation.py

```diff
@@ -0,0 +1,1120 @@
+                 \cdots \right].
+
+        *This is an optional method.* It is needed when transformation of
+        standard deviation :meth:`Transformation.convert_standard_deviation` or
+        covariance matrix :meth:`Transformation.convert_covariance_matrix` is
+        needed, or when ``evaluateS1()`` is needed.
+        """
+        raise NotImplementedError
+
+    def jacobian_S1(self, q):
+        r"""
+        Computes the Jacobian matrix of the transformation calculated at the
+        parameter vector ``q`` in the search space, and returns the result
+        along with the partial derivatives of the result with respect to the
+        parameters.
+
+        The returned data is a tuple ``(S, S')`` where ``S`` is a
+        ``n_parameters`` by ``n_parameters`` matrix and ``S'`` is a sequence of
+        ``n_parameters`` matrices.
+
+        *This is an optional method.* It is needed when the transformation is
+        used along with a non-element-wise transformation in
+        :class:`ComposedTransformation`.
+        """
+        raise NotImplementedError
+
+    def log_jacobian_det(self, q):
+        """
+        Returns the logarithm of the absolute value of the determinant of the
+        Jacobian matrix of the transformation :meth:`Transformation.jacobian`
+        calculated at the parameter vector ``q`` in the search space.
+
+        *This is an optional method.* It is needed when transformation is
+        performed on :class:`LogPDF` and/or that requires ``evaluateS1()``;
+        e.g. not necessary if it's used for :class:`ErrorMeasure` without
+        :meth:`ErrorMeasure.evaluateS1()`.
+        """
+        return np.log(np.abs(np.linalg.det(self.jacobian(q))))
+
+    def log_jacobian_det_S1(self, q):
+        """
+        Computes the logarithm of the absolute value of the determinant of the
+        Jacobian, and returns the result plus the partial derivatives of the
+        result with respect to the parameters.
+
+        The returned data is a tuple ``(S, S')`` where ``S`` is a scalar value
+        and ``S'`` is a sequence of length ``n_parameters``.
+
+        Note that the derivative returned is of the log of the determinant of
+        the Jacobian, so ``S' = d/dq log(|det(J(q))|)``, evaluated at input.
+
+        *This is an optional method.* It is needed when transformation is
+        performed on :class:`LogPDF` and that requires ``evaluateS1()``.
+        """
+        # Directly calculate the derivative using jacobian_S1()
+        #
+        # d/dq log(|det(J(q))|) = 1/|det(J(q))| * sign(det(J(q)))
+        #                         * d/dq det(J(q))
+        #
+        # The second term is given by Eq. 46 in the Matrix Cookbook (2012)
+        # http://www.math.uwaterloo.ca/~hwolkowi/matrixcookbook.pdf
+        #
+        # d/dq det(J(q)) = det(J) Tr(J^{-1} d/dq J)
+        #
+        # Therefore
+        #
+        # d/dq log(|det(J(q))|) = Tr(J^{-1} d/dq J)
+        #
+        q = pints.vector(q)
```

Comment:
Same here. And the default implementation could be explained in the docstring?

Reply:
Added explanation
```
        The absolute value of the determinant of the Jacobian is provided by    
        :meth:`Transformation.log_jacobian_det`. The default implementation     
        calculates the derivatives of the log-determinant using [1]_            
                                                                                
        .. math::                                                               
            \frac{d}{dq} \log(|det(\mathbf{J})|) =                              
                trace(\mathbf{J}^{-1} \frac{d}{dq} \mathbf{J}),                 
                                                                                
        where the derivative of the Jacobian matrix is provided by              
        :meth:`Transformation.jacobian_S1` and the matrix inversion is          
        numerically calculated. If there is an analytic expression for the      
        specific transformation, a reimplementation of this method may be       
        preferred.
```
Any suggestion?